### PR TITLE
Tentative solution to whitelisting wildcards with subdomains logic.

### PR DESCRIPTION
**Pull request by tashachin**
Resolves https://github.com/EFForg/privacybadger/issues/1519

I've written a few unit tests for edge cases and got a working solution. After spending some time trying to refactor my code, I can't figure out how to make my inequality checks DRYer.

I have three cases: If the wildcard's num of subdomains matches the origin URL, if it is longer, or if it is shorter. I can't think of a way to simplify it because of how subtraction is not commutative (the order in which it is performed matters).

Any feedback is greatly appreciated!

### CODE REVIEW

**Comment by ghostwords** [comment]
Looks like "addresses" isn't one of [GitHub's auto-issue-closing keywords](https://help.github.com/articles/closing-issues-using-keywords/). You can tell the keyword was understood when it gets a help tooltip underline.

**Comment by ghostwords** [comment]
I added some linting notes to docs in 8eb8b8ef2b6073123d9a0434e784d19e79a116ff.

**Comment by tashachin** [comment]
@ghostwords Thanks for the feedback!

I'm having some trouble with Github at the moment, but I'll resolve the linting issues over the weekend. (Just wanted to keep you posted on my timeframe).

File: src/js/background.js [hunk 1]
```diff
@@ -817,4 +821,4 @@ startBackgroundListeners();
 
 // TODO move listeners and this message behind INITIALIZED
 console.log('Privacy badger is ready to rock!');
-console.log('Set DEBUG=1 to view console messages.');
+console.log('Set DEBUG=1 to view console messages.');
```
Comment (ghostwords) [inline src/js/background.js]: Style nit: missing end-of-file newline. This change shouldn't be here, basically.

File: src/tests/tests/utils.js [hunk 2]
```diff
@@ -65,6 +65,17 @@
     assert.ok(!badger.isPrivacyBadgerEnabled("example.com"), "disabled for site");
     badger.enablePrivacyBadgerForOrigin("example.com");
     assert.ok(badger.isPrivacyBadgerEnabled("example.com"), "enabled for site");
+
+    // whitelist wildcard cases
+    badger.disablePrivacyBadgerForOrigin('*.mail.example.com');
+    assert.ok(badger.isPrivacyBadgerEnabled('www.example.com'), "enabled for site");
+    assert.ok(badger.isPrivacyBadgerEnabled('web.stuff.example.com'), "enabled for site");
+    assert.ok(!badger.isPrivacyBadgerEnabled('web.mail.example.com'), "disabled for site");
```
Comment (ghostwords) [inline src/tests/tests/utils.js]: We could use a more specific [QUnit assertion](https://api.qunitjs.com/assert/) here (and below): `assert.notOk`.

File: src/tests/tests/utils.js [hunk 3]
```diff
@@ -65,6 +65,17 @@
     assert.ok(!badger.isPrivacyBadgerEnabled("example.com"), "disabled for site");
     badger.enablePrivacyBadgerForOrigin("example.com");
     assert.ok(badger.isPrivacyBadgerEnabled("example.com"), "enabled for site");
+
+    // whitelist wildcard cases
+    badger.disablePrivacyBadgerForOrigin('*.mail.example.com');
+    assert.ok(badger.isPrivacyBadgerEnabled('www.example.com'), "enabled for site");
```
Comment (ghostwords) [inline src/tests/tests/utils.js]: Would be nice to provide more descriptive messages. What are we asserting with each statement?

File: src/js/background.js [hunk 4]
```diff
@@ -592,11 +592,15 @@ Badger.prototype = {
     if (disabledSites && disabledSites.length > 0) {
       for (var i = 0; i < disabledSites.length; i++) {
         var site = disabledSites[i];
+
         if (site.startsWith("*")) {
-          if (window.getBaseDomain(site) === window.getBaseDomain(origin)) {
+          var wildcard = site.slice(2); // remove "*."
```
Comment (ghostwords) [inline src/js/background.js]: Should this be `site.slice(1)` instead? If we remove two characters, `*.example.com` will match `example.com`.

Also, the dot may not be there, which is fine if we remove one character (but not quite right if we remove two).

File: src/js/background.js [hunk 5]
```diff
@@ -592,11 +592,15 @@ Badger.prototype = {
...
+
```
Comment (ghostwords) [inline src/js/background.js]: Style nit: Could this be a blank line (instead of a spaces-only line)?

File: src/js/background.js (hunk 4 above)
Comment (ghostwords) [reply src/js/background.js]: Would be good to add a test for this scenario.

File: src/tests/tests/utils.js [hunk 6]
```diff
@@ -65,6 +65,17 @@
...
+    assert.ok(badger.isPrivacyBadgerEnabled('web.stuff.example.com'), "enabled for site");
+    assert.ok(!badger.isPrivacyBadgerEnabled('web.mail.example.com'), "disabled for site");
+
+    badger.disablePrivacyBadgerForOrigin('*.fakedomain.com');
```
Comment (ghostwords) [inline src/tests/tests/utils.js]: What's the difference between this set of assertions and the one above?

**Comment by ghostwords** [comment]
Let me know if you'd like me to finish this PR. This is close to complete! Either way, thank you for your contribution!

**Comment by tashachin** [comment]
Apologies for the wait! I won't be able to review your feedback until this Sunday at the soonest, but if you'd like to finish it, by all means, don't wait for me!

**Comment by ghostwords** [comment]
That's fine, no rush, let's finish it together.

File: src/js/background.js (hunk 1 above)
Comment (tashachin) [reply src/js/background.js]: Hey Alexei,

Would you prefer that I remove this change from the original commit OR make a new commit to remove the change? (Kai's right beside me telling me my editor added the change without my knowledge).

Comment (ghostwords) [reply src/js/background.js]: Probably more straightforward to do followup commits (rather than rewrite history).

(I'm about to step out, won't be able to respond quickly, sorry.)

**Comment by ghostwords** [comment]
I added another test in 6817417f9c0f6302b2c342eadd620dec6d4ca7ee.
